perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_user_attribute_def_virt_kerberosLogins.java

```diff
@@ -64,27 +65,31 @@ public Attribute getAttributeValue(PerunSessionImpl sess, User user, AttributeDe
 	}
 
 	@Override
-	public List<String> resolveVirtualAttributeValueChange(PerunSessionImpl perunSession, String message) throws InternalErrorException, WrongReferenceAttributeValueException, AttributeNotExistsException, WrongAttributeAssignmentException {
-		List<String> resolvingMessages = new ArrayList<String>();
+	public List<AuditEvent> resolveVirtualAttributeValueChange(PerunSessionImpl perunSession, AuditEvent message) throws InternalErrorException, WrongReferenceAttributeValueException, AttributeNotExistsException, WrongAttributeAssignmentException {
+		List<AuditEvent> resolvingMessages = new ArrayList<>();
 		if (message == null) return resolvingMessages;
 
-		Matcher addUserExtSourceMatcher = addUserExtSource.matcher(message);
-		Matcher removeUserExtSourceMatcher = removeUserExtSource.matcher(message);
-		Matcher extSourceKerberosMatcher = extSourceKerberos.matcher(message);
+		if (message instanceof UserExtSourceAddedToUser
+			&& ((UserExtSourceAddedToUser) message).getUserExtSource().getExtSource() instanceof ExtSourceKerberos) {
 
-		User user = null;
-		Attribute attrVirtKerberosLogins = null;
+			resolvingMessages.addAll(resolveEvent(perunSession, ((UserExtSourceAddedToUser) message).getUser()));
+		}
 
-		if(extSourceKerberosMatcher.find()) {
-			if (addUserExtSourceMatcher.find() || removeUserExtSourceMatcher.find()) {
-				user = perunSession.getPerunBl().getModulesUtilsBl().getUserFromMessage(perunSession, message);
-				if (user != null) {
-					attrVirtKerberosLogins = perunSession.getPerunBl().getAttributesManagerBl().getAttribute(perunSession, user, AttributesManager.NS_USER_ATTR_VIRT + ":kerberosLogins");
-					String messageAttributeSet = attrVirtKerberosLogins.serializeToString() + " set for " + user.serializeToString() + ".";
-					resolvingMessages.add(messageAttributeSet);
-				}
-			}
+		if (message instanceof UserExtSourceRemovedFromUser
+			&& ((UserExtSourceRemovedFromUser) message).getUserExtSource().getExtSource() instanceof ExtSourceKerberos) {
+
+			resolvingMessages.addAll(resolveEvent(perunSession, ((UserExtSourceRemovedFromUser) message).getUser()));
 		}
+
+		return resolvingMessages;
+	}
+
+	private List<AuditEvent> resolveEvent(PerunSessionImpl perunSession, User user) throws WrongAttributeAssignmentException, InternalErrorException, AttributeNotExistsException {
+		List<AuditEvent> resolvingMessages = new ArrayList<>();
+
+		Attribute attrVirtKerberosLogins = perunSession.getPerunBl().getAttributesManagerBl().getAttribute(perunSession, user, A_U_V_KERBEROS_LOGINS);
```

Comment:
Please, add here this comment: `//FIXME: if the attribute value is null or empty, this method should return AttributeRemovedForUser instead`